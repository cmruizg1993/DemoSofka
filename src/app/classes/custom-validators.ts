import { AbstractControl, AsyncValidatorFn, ValidationErrors, ValidatorFn } from "@angular/forms";
import { ProductService } from "../services/product.service";
import { Observable, map } from "rxjs";

export  class CustomValidators {
    static productIDValidator(productService: ProductService): AsyncValidatorFn {
        return (control: AbstractControl): Observable<ValidationErrors | null> => {
            const id = (control.value as string).trim().toLowerCase();
            const response$ = productService.verificationID(id);
            return response$.pipe(map(isExisting => (isExisting ? { idExists: true } : null)));
        };
    }
    static productValidateDateFormat(): ValidatorFn{
        return (control: AbstractControl): ValidationErrors | null => {
            const regex = /[0-9]{4,4}-[0-9]{2,2}-[0-9]{2,2}/;
            const date = new Date(control.value);
            return !regex.test(control.value) || isNaN(date.getDate()) ? { invalidDateFormat: { value: control.value } } : null;
        };
    }
    static productValidateDateValue(): ValidatorFn{
        
        return (control: AbstractControl): ValidationErrors | null => {
            const date = new Date(control.value);
            const currentDate = new Date();
            return date < currentDate ? { invalidDateValue: { value: control.value } } : null;
        };
    }
}
