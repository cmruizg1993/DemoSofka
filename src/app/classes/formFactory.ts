
import { IForm } from "../interfaces/i-form";
import { IFormField } from "../interfaces/i-form-field";

export abstract class FormFactory implements IForm{
  
  formFields: IFormField[] = [];

  getFormField(name: string): IFormField {
    let field = this.formFields.find( f => f.name == name );
    
    if(!field)
      field = {
        name
      };
    return field;
  }
  getCreationFormFields(): IFormField[] {
    throw new Error("Method not implemented.");
  }
  getUpdateFormFields(): IFormField[] {
    throw new Error("Method not implemented.");
  }
    
}
