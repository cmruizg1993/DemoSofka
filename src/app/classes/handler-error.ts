import { Observable, throwError } from "rxjs";
import { IError, IHandlerError } from "../interfaces/i-handler-error";
import { HttpErrorResponse } from "@angular/common/http";

export class HandlerError implements IHandlerError{
    
    handlerError(errorResponse: HttpErrorResponse): Observable<never> {
        console.log('OcurriÃ³ un error: ', errorResponse);
        return throwError(()=>{
            const error: IError ={
                name: errorResponse.name,
                message: errorResponse.message
            }
            if(!!errorResponse.error?.name)
                error.name = errorResponse.error.name;
            if(!!errorResponse.error?.message)
                error.message = errorResponse.error.message;
            if(!!errorResponse.error?.stack)
                error.stack = errorResponse.error.stack;
            return error;
        })
    }

}