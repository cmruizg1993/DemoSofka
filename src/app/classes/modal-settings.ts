import { Observable, Subject } from "rxjs";

export class ModalSettings {
    title: string = '';
    content: string = '';
    confirmButton = false;
    cancelButton = false;
    confirmButtonLabel: string = '';
    cancelButtonLabel: string = '';
    
    confirmAction?: CallableFunction;
    cancelAction?: CallableFunction; 

    showModal$ = new Subject<boolean>();
    get show$(): Observable<boolean>{
        return this.showModal$.asObservable();
    }
    open(){
        this.showModal$.next(true)
    }
    close(){
        this.showModal$.next(false);
    }
}
