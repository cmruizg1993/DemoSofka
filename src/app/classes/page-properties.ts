export class PagerProperties{
    page: number = 1;
    pageSize = 5;
  }