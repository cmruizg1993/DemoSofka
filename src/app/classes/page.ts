export class Page{
    number: number = 0;
    isActive: boolean = false;
  }