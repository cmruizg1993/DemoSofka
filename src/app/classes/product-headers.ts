import { IHeader } from "../interfaces/i-header";

export class ProductTableInfo{
    public static getHeaders(): IHeader[] {
        return [
            { key: 'id', label: 'Id', hidden: true },
            { key: 'logo', label: 'Logo', isImage: true },
            { key: 'name', label: 'Nombre del prducto' },
            { key: 'description', label: 'Descripcion', info: 'Representa a descripción del producto financiero' },
            { key: 'date_release', label: 'Fecha de liberación', info: 'Representa la fecha de liberación del producto financiero' },
            { key: 'date_revision', label: 'Fecha de reestructuración',  info: 'Representa la fecha de revisión del producto financiero' } ];
    }
}