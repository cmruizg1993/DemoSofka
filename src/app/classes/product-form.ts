import { FormGroup } from "@angular/forms";
import { IForm } from "../interfaces/i-form";
import { IFormField } from "../interfaces/i-form-field";
import { ProductService } from "../services/product.service";
import { CustomValidators } from "./custom-validators";
import { Product } from "./product";
import { FormFactory } from "./formFactory";

export class ProductForm extends FormFactory{


    private productService: ProductService|null

    private product: Product|null;

    constructor(product: Product|null, productService: ProductService|null){
        super();
        this.productService = productService;
        this.product = product;
    }

    override getCreationFormFields(): IFormField[] {
        this.resetProductFormFields();
        if(!this.productService)
            throw new Error("Illegal invocation, productService not provided.");         
        this.getFormField("id")
                ?.asynValidators?.push(CustomValidators.productIDValidator(this.productService)); 
        console.log(this.getFormField("date_release"));  
        return this.formFields;
    }
    override getUpdateFormFields(): IFormField[] {        
        if(!this.product)
            throw new Error("Illegal invocation, product Object not provided.");
        this.resetProductFormFields();
        this.getFormField("id").value = this.product.id;
        this.getFormField("id").disabled = true;
        this.getFormField("name").value = this.product.name;
        this.getFormField("description").value = this.product.description;
        this.getFormField("logo").value = this.product.logo;
        this.getFormField("date_release").value = this.product.date_release;        
        this.getFormField("date_revision").value = this.product.date_revision;
        
        return this.formFields;
    }
    private resetProductFormFields(){
        this.formFields = [
            {
              name: 'id',
              label: 'ID',
              required: true,
              pattern: '^[A-Za-z0-9]{3,10}$',
              asynValidators: []
            },
            {
              name: 'name',
              label: 'Nombre',
              required: true,
              pattern: '^[A-Za-z0-9 áéíóúÁÉÍÓÚñÑ]{5,100}$'
            },
            {
              name: 'description',
              label: 'Descripción',
              required: true,
              pattern: '^[A-Za-z0-9 áéíóúÁÉÍÓÚñÑ]{10,100}$',
              female: true
            },
            {
              name: 'logo',
              label: 'Logo',
              required: true
            },
            {
              name: 'date_release',
              label: 'Fecha liberación',
              placeholder: 'yyyy-mm-dd',
              required: true,
              pattern: '^20[0-9]{2,2}-[01]{1,1}[0-9]{1,1}-[0123]{1,1}[0-9]{1,1}$',
              female: true,
              validators: [CustomValidators.productValidateDateFormat(), CustomValidators.productValidateDateValue()],
              onValidation: (value: string, form: FormGroup) =>
                { 
                  const control = form.get('date_revision');
                  const releaseDate = value.split('-');
                  control?.setValue(`${Number(releaseDate[0])+1}-${releaseDate[1]}-${releaseDate[2]}`);          
                }
            },
            {
              name: 'date_revision',
              label: 'Fecha revisión',
              disabled: true,
              required: true
            }
          ];
    }
}
