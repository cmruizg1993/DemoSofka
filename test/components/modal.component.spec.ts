import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { ModalComponent } from '../../src/app/components/modal/modal.component';
import { ModalSettings } from '../../src/app/classes/modal-settings';



describe('AppComponent', () => {
  let fixture: ComponentFixture<ModalComponent>;
  let compiled: HTMLElement;
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        RouterTestingModule
      ],
      declarations: [
        ModalComponent
      ],
    }).compileComponents();
    fixture = TestBed.createComponent(ModalComponent);
    fixture.componentInstance.modalSettings = new ModalSettings();
  
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('Se debe crear el componente', () => {
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });
  it(`Debe coincidir con el snapshot'`, () => {
    expect(compiled).toMatchSnapshot();
  });
})