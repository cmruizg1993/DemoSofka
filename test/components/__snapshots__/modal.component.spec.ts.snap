// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent Debe coincidir con el snapshot' 1`] = `
<div
  id="root1"
>
  <div
    class="modal-backdrop"
  >
    <div
      class="modal"
    >
      
      
    </div>
  </div>
</div>
`;
